File: Data Structures/Assignment4/is_bst.java
import java.util.*;
import java.io.*;

public class is_bst {
    class FastScanner {
        StringTokenizer tok = new StringTokenizer("");
        BufferedReader in;
        
        FastScanner() {
            in = new BufferedReader(new InputStreamReader(System.in));
        }
        
        String next() throws IOException {
            while (!tok.hasMoreElements())
                tok = new StringTokenizer(in.readLine());
            return tok.nextToken();
        }
        int nextInt() throws IOException {
            return Integer.parseInt(next());
        }
    }
    
    public class IsBST {
        class Node {
            int key;
            int left;
            int right;
            
            Node(int key, int left, int right) {
                this.left = left;
                this.right = right;
                this.key = key;
            }
        }
        private ArrayList<Integer> res = new ArrayList<>();
        int nodes;
        Node[] tree;
        
        void read() throws IOException {
            FastScanner in = new FastScanner();
            nodes = in.nextInt();
            tree = new Node[nodes];
            for (int i = 0; i < nodes; i++) {
                tree[i] = new Node(in.nextInt(), in.nextInt(), in.nextInt());
            }
        }
        
        void inOrderTraverse(int i) {
            if (tree[i].left != -1) {
                inOrderTraverse(tree[i].left);
            }
            res.add(tree[i].key);
            if (tree[i].right != -1) {
                inOrderTraverse(tree[i].right);;
            }
        }
        boolean solve() {
            res.clear();
            if (tree.length <= 1) {
                return true;
            }
            inOrderTraverse(0);
            for (int i = 0; i < res.size() - 1; i++) {
                if (res.get(i) > res.get(i + 1)) {
                    return false;
                }
            }
            return true;
        }
    }
    
    static public void main(String[] args) throws IOException {
        new Thread(null, new Runnable() {
            public void run() {
                try {
                    new is_bst().run();
                } catch (IOException e) {
                }
            }
        }, "1", 1 << 26).start();
    }
    public void run() throws IOException {
        IsBST tree = new IsBST();
        tree.read();
        if (tree.solve()) {
            System.out.println("CORRECT");
        } else {
            System.out.println("INCORRECT");
        }
    }
}

File: Data Structures/Assignment4/is_bst.cpp
#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

struct Node {
    int key;
    int left;
    int right;
    
    Node() : key(0), left(-1), right(-1) {}
    Node(int key_, int left_, int right_) : key(key_), left(left_), right(right_) {}
};

class Helper {
private:
    vector<int> res;
    vector<Node>& Tree;
public:
    Helper(vector<Node>& tree) : Tree(tree) {}
    void inOrderTraverse(int i) {
        if (Tree[i].left != -1) {
            inOrderTraverse(Tree[i].left);
        }
        res.push_back(Tree[i].key);
        if (Tree[i].right != -1) {
            inOrderTraverse(Tree[i].right);
        }
    }
    bool IsBinarySearchTree() {
        res.clear();
        inOrderTraverse(0);
        for (int i = 0; i < (int)res.size() - 1; i++) {
            if (res[i] > res[i + 1]) {
                return false;
            }
        }
        return true;
    }
};



int main() {
    int nodes;
    cin >> nodes;
    if (nodes <= 1 && nodes >= 0) {
        cout << "CORRECT";
        return 0;
    }
    vector<Node> tree;
    for (int i = 0; i < nodes; ++i) {
        int key, left, right;
        cin >> key >> left >> right;
        tree.push_back(Node(key, left, right));
    }
    Helper helper = Helper(tree);
    if (helper.IsBinarySearchTree()) {
        cout << "CORRECT" << endl;
    } else {
        cout << "INCORRECT" << endl;
    }
    return 0;
}
